pkg/virt-controller/services/rendervolumes.go

```diff
@@ -0,0 +1,572 @@
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/config"
+	"kubevirt.io/kubevirt/pkg/hooks"
+	hostdisk "kubevirt.io/kubevirt/pkg/host-disk"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/util/types"
+)
+
+type VolumeRendererOption func(renderer *VolumeRenderer) error
+
+type VolumeRenderer struct {
+	containerDiskDir string
+	ephemeralDiskDir string
+	virtShareDir     string
+	namespace        string
+	vmiVolumes       []v1.Volume
+	podVolumes       []k8sv1.Volume
+	podVolumeMounts  []k8sv1.VolumeMount
+	volumeDevices    []k8sv1.VolumeDevice
+}
+
+func NewVolumeRenderer(namespace string, ephemeralDisk string, containerDiskDir string, virtShareDir string, volumeOptions ...VolumeRendererOption) (*VolumeRenderer, error) {
+	volumeRenderer := &VolumeRenderer{
+		containerDiskDir: containerDiskDir,
+		ephemeralDiskDir: ephemeralDisk,
+		namespace:        namespace,
+		virtShareDir:     virtShareDir,
+	}
+	for _, volumeOption := range volumeOptions {
+		if err := volumeOption(volumeRenderer); err != nil {
+			return nil, err
+		}
+	}
+	return volumeRenderer, nil
+}
+
+func (vr *VolumeRenderer) Mounts() []k8sv1.VolumeMount {
+	volumeMounts := []k8sv1.VolumeMount{
+		mountPath("private", util.VirtPrivateDir),
+		mountPath("public", util.VirtShareDir),
+		mountPath("ephemeral-disks", vr.ephemeralDiskDir),
+		mountPathWithPropagation(containerDisks, vr.containerDiskDir, k8sv1.MountPropagationHostToContainer),
+		mountPath("libvirt-runtime", "/var/run/libvirt"),
+		mountPath("sockets", filepath.Join(vr.virtShareDir, "sockets")),
+	}
+	return append(volumeMounts, vr.podVolumeMounts...)
+}
+
+func (vr *VolumeRenderer) Volumes() []k8sv1.Volume {
+	volumes := []k8sv1.Volume{
+		emptyDirVolume("private"),
+		emptyDirVolume("public"),
+		emptyDirVolume("sockets"),
+		emptyDirVolume(virtBinDir),
+		emptyDirVolume("libvirt-runtime"),
+		emptyDirVolume("ephemeral-disks"),
+		emptyDirVolume(containerDisks),
+	}
+	return append(volumes, vr.podVolumes...)
+}
+
+func (vr *VolumeRenderer) VolumeDevices() []k8sv1.VolumeDevice {
+	return vr.volumeDevices
+}
+
+func mountPath(name string, path string) k8sv1.VolumeMount {
+	return k8sv1.VolumeMount{
+		Name:      name,
+		MountPath: path,
+	}
+}
+
+func mountPathWithPropagation(name string, path string, propagation k8sv1.MountPropagationMode) k8sv1.VolumeMount {
+	return k8sv1.VolumeMount{
+		Name:             name,
+		MountPath:        path,
+		MountPropagation: &propagation,
+	}
+}
+
+func emptyDirVolume(name string) k8sv1.Volume {
+	return k8sv1.Volume{
+		Name: name,
+		VolumeSource: k8sv1.VolumeSource{
+			EmptyDir: &k8sv1.EmptyDirVolumeSource{}},
+	}
+}
+
+func withVMIVolumes(pvcStore cache.Store, vmiSpecVolumes []v1.Volume, vmiVolumeStatus []v1.VolumeStatus) VolumeRendererOption {
+	return func(renderer *VolumeRenderer) error {
+		for _, volume := range vmiSpecVolumes {
+			if _, isHotplugVolume := hotplugVolumes(vmiVolumeStatus, vmiSpecVolumes)[volume.Name]; isHotplugVolume {
```

Comment:
I do not understand why the creation of `hotplugVolumes` map is inlined here. It now has square time complexity and very bad memory behavior.  For each volume, it creates new hotplugVolumes to just check if the volume is of type hotplug.

```
hotplugVolumesByName := hotplugVolumes(vmiVolumeStatus, vmiSpecVolumes)
			if _, isHotplugVolume := hotplugVolumesByName[volume.Name]; isHotplugVolume {
			
```

Reply:
Good catch. Looking at it now, I have no clue why I wrote it like that.

Reply:
Done.